### PR TITLE
feat(datasets) Add label count utils

### PR DESCRIPTION
**Pull request by adam-narozniak**
## Issue
Divide up the https://github.com/adap/flower/pull/3451 PR

## Proposal
Introduce the public methods for `compute_counts` and `compute_frequencies`.

### CODE REVIEW

File: datasets/flwr_datasets/metrics/utils.py [hunk 1]
```diff
@@ -13,14 +13,189 @@
...
+def compute_frequencies(
+    partitioner: Partitioner,
+    column_name: str,
+    verbose_names: bool = False,
+    max_num_partitions: Optional[int] = None,
+) -> pd.DataFrame:
+    """Compute the frequencies of unique values in a given column in the partitions.
+
+    The frequencies sum up to 1 for a given partition id. Take into account all
+    possible labels in dataset when computing count for each partition (assign 0 as the
```
Comment (jafermarq) [inline datasets/flwr_datasets/metrics/utils.py:143]:
```suggestion
    The frequencies sum up to 1 for a given partition id. This function takes into account all
    possible labels in the dataset when computing the count for each partition (assign 0 as the
```

File: datasets/flwr_datasets/metrics/utils.py [hunk 2]
```diff
@@ -13,14 +13,189 @@
...
+        feature of type `ClassLabel`.
+    max_num_partitions : Optional[int]
+        The number of partitions that will be used. If left None, then all partitions
+        will be used.
+
+    Returns
+    -------
+    dataframe: pd.DataFrame
+        DataFrame where the rows represent the partition id and the column represent
+        the unique values found in column specified by `column_name`.
```
Comment (jafermarq) [inline datasets/flwr_datasets/metrics/utils.py:164]: needs to be updated to mention values are "ratios"?

File: datasets/flwr_datasets/metrics/utils.py [hunk 3]
```diff
@@ -13,14 +13,189 @@
...
+        Partitioner with an assigned dataset.
+    column_name : str
+        Column name identifying label based on which the count will be calculated.
+    verbose_names : bool
+        Whether to use verbose versions of the values in the column specified by
+        `column_name`. The verbose value are possible to extract if the column is a
+        feature of type `ClassLabel`.
+    max_num_partitions : Optional[int]
+        The number of partitions that will be used. If left None, then all partitions
+        will be used.
```
Comment (jafermarq) [inline datasets/flwr_datasets/metrics/utils.py:158]: Similarly to the comment for `compute_count`, should we indicate that the first N are returned ?

File: datasets/flwr_datasets/metrics/utils_test.py [hunk 4]
```diff
@@ -19,12 +19,126 @@
...
+            ),
+        )
+        iid_partitioner.dataset = dataset
+        result = self.result.copy()
+        result.columns = new_col_names
+        result = result.div(result.sum(axis=1), axis=0)
+        frequencies = compute_frequencies(
+            iid_partitioner, column_name="label", verbose_names=True
+        )
+        print(frequencies)
```
Comment (jafermarq) [inline datasets/flwr_datasets/metrics/utils_test.py:119]: remove?

File: datasets/flwr_datasets/metrics/utils.py [hunk 5]
```diff
@@ -13,14 +13,189 @@
...
+
+    Returns
+    -------
+    dataframe: pd.DataFrame
+        DataFrame where the rows represent the partition id and the column represent
+        the unique values found in column specified by `column_name`.
```
[the remark this answers is not in the document]
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils.py:54]: Actually, there I meant that the columns' values are unique and created from `column_name` column.  What I tried to say that the list of columns is e.g. [1,2, ... 9]. I haven't said anything about the values of the dataframe (let me also add it now).
So maybe I should say raw index and column index when describing that.

File: datasets/flwr_datasets/metrics/utils.py [hunk 6]
```diff
@@ -13,14 +13,189 @@
...
+            f"The specified 'column_name': '{column_name}' is not present in the "
+            f"dataset. The dataset contains columns {partitioner.dataset.column_names}."
+        )
+
+    if max_num_partitions is None:
+        max_num_partitions = partitioner.num_partitions
+    else:
+        max_num_partitions = min(max_num_partitions, partitioner.num_partitions)
+    assert isinstance(max_num_partitions, int)
+    partitions = [partitioner.load_partition(i) for i in range(max_num_partitions)]
```
[the remark this answers is not in the document]
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils.py:92]: Memory-wise, it's not a problem but I'll switch to what you're proposing

File: datasets/flwr_datasets/metrics/utils_test.py (hunk 4 above)
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils_test.py:119]: Yes, sorry, I was debugging and forgot about it

File: datasets/flwr_datasets/metrics/utils.py (hunk 2 above)
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils.py:164]: Rewritten

File: datasets/flwr_datasets/metrics/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils.py:54]: Change it, lemme wdyt?

File: datasets/flwr_datasets/metrics/utils.py [hunk 7]
```diff
@@ -13,14 +13,189 @@
...
+    partitioner : Partitioner
+        Partitioner with an assigned dataset.
+    column_name : str
+        Column name identifying label based on which the count will be calculated.
+    verbose_names : bool
+        Whether to use verbose versions of the values in the column specified by
+        `column_name`. The verbose value are possible to extract if the column is a
+        feature of type `ClassLabel`.
+    max_num_partitions : Optional[int]
+        The number of partitions that will be used. If left None, then all partitions
```
[the remark this answers is not in the document]
Comment (adam-narozniak) [reply datasets/flwr_datasets/metrics/utils.py:47]: Rewritten now

File: datasets/flwr_datasets/metrics/utils.py [hunk 8]
```diff
@@ -13,14 +13,197 @@
...
+        The maximum number of partitions that will be used. If greater than the
+        total number of partitions in a partitioner, it won't have an effect. If left
+        as None, then all partitions will be used.
+
+    Returns
+    -------
+    dataframe: pd.DataFrame
+        DataFrame where the row index represent the partition id and the column index
+        represent the unique values found in column specified by `column_name`
+        (e.g. represeting the labels). The value of the dataframe.loc[i, j] represnt
```
Comment (jafermarq) [inline datasets/flwr_datasets/metrics/utils.py:56]:
```suggestion
        (e.g. represeting the labels). The value of the dataframe.loc[i, j] represents
```

File: datasets/flwr_datasets/metrics/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jafermarq) [reply datasets/flwr_datasets/metrics/utils.py]: looks much clearer than my suggestion

**Review by jafermarq — APPROVED** [review]

**Pull request merged**